### PR TITLE
[SPARK-19561][SQL] add int case handling for TimestampType

### PR DESCRIPTION
**Pull request by JasonMWhite**
## What changes were proposed in this pull request?

Add handling of input of type `Int` for dataType `TimestampType` to `EvaluatePython.scala`. Py4J serializes ints smaller than MIN_INT or larger than MAX_INT to Long, which are handled correctly already, but values between MIN_INT and MAX_INT are serialized to Int.

These range limits correspond to roughly half an hour on either side of the epoch. As a result, PySpark doesn't allow TimestampType values to be created in this range.

Alternatives attempted: patching the `TimestampType.toInternal` function to cast return values to `long`, so Py4J would always serialize them to Scala Long. Python3 does not have a `long` type, so this approach failed on Python3.

## How was this patch tested?

Added a new PySpark-side test that fails without the change.

The contribution is my original work and I license the work to the project under the project’s open source license.

Resubmission of https://github.com/apache/spark/pull/16896. The original PR didn't go through Jenkins and broke the build. @davies @dongjoon-hyun

@cloud-fan Could you kick off a Jenkins run for me? It passed everything for me locally, but it's possible something has changed in the last few weeks.

### CODE REVIEW

**Comment by cloud-fan** [comment]
ok to test

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -1435,6 +1435,12 @@ def test_time_with_timezone(self):
         self.assertEqual(now, now1)
         self.assertEqual(now, utcnow1)
 
+    # regression test for SPARK-19561
+    def test_datetime_at_epoch(self):
+        epoch = datetime.datetime.fromtimestamp(0)
+        df = self.spark.createDataFrame([Row(date=epoch)])
+        self.assertEqual(df.first()['date'], epoch)
```
Comment (viirya) [inline python/pyspark/sql/tests.py]: This test is invalid. `df.first()['date']` is `None` even in current master branch.

Comment (JasonMWhite) [inline python/pyspark/sql/tests.py]: Yes, that's the bug this PR is fixing. It shouldn't be `None`.

**Comment by JasonMWhite** [comment]
Ah, test failed on Python 3.4 only. That makes some sense, I only tested locally on 2.6, and there are changes with how Python 3 handles ints vs longs. I'll dig in with Python 3.4 and see if I can see the cause for the test failure.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -189,7 +189,7 @@ def toInternal(self, dt):
         if dt is not None:
             seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
```
Comment (viirya) [inline python/pyspark/sql/types.py]: hmm, for the value of `epoch = datetime.datetime.fromtimestamp(0)`, `seconds` is 0. What is it different to use `int` or `long`?

Comment (JasonMWhite) [inline python/pyspark/sql/types.py]: The JIRA ticket has the details: https://issues.apache.org/jira/browse/SPARK-19561. But in a nutshell, that's the point: `int(0)` fails but `long(0)` succeeds.

Comment (JasonMWhite) [inline python/pyspark/sql/types.py]: https://github.com/bartdag/py4j/blob/master/py4j-python/src/py4j/protocol.py#L271-L275

Py4J automatically serializes any Python integer larger than 2 ^ 31 as `LONG_TYPE`, otherwise it's `INTEGER_TYPE`. Python longs are always serialized as `LONG_TYPE`.

I suspect my issue with Python 3 is that there is no more `long`, it's all just `int`. This may require a fix on the Scala side to accept either an int or a long to the appropriate constructor.

Comment (viirya) [inline python/pyspark/sql/types.py]: Thanks. Interesting.

Comment (viirya) [inline python/pyspark/sql/types.py]: I just tested it. In Python3, even `toInternal` returns Python's `long`, you still can a `java.lang.Integer` in JVM side.

However, in Python2, you can get `java.lang.Long`.

Comment (viirya) [inline python/pyspark/sql/types.py]: Due the Python3 doesn't have `long` anymore, I think we can't solve this in python. I can fix this in JVM side.

Comment (JasonMWhite) [inline python/pyspark/sql/types.py]: Agreed.

Comment (viirya) [inline python/pyspark/sql/types.py]: @JasonMWhite Are you going to submit another PR for it?

Comment (JasonMWhite) [inline python/pyspark/sql/types.py]: I will, yes. Trying to find where the appropriate location is in the Scala code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala [hunk 3]
```diff
@@ -112,6 +112,7 @@ object EvaluatePython {
     case (c: Int, DateType) => c
 
     case (c: Long, TimestampType) => c
+    case (c: Int, TimestampType) => c.toLong
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala]: Can you add a comment for the reason?

**Comment by viirya** [comment]
Btw, since it is a change to SQL code. Better to add `[SQL]` to the title.

**Comment by viirya** [comment]
LGTM except for a minor comment.

**Comment by viirya** [comment]
oh, the PR description is not correct now. Can you update it too?

**Comment by cloud-fan** [comment]
LGTM

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.1!

**Comment by JasonMWhite** [comment]
Thanks @cloud-fan !
